operator/standalone/update.go

```diff
@@ -0,0 +1,386 @@
+
+import (
+	"context"
+	"fmt"
+	pipeline "github.com/ccremer/go-command-pipeline"
+	helmv1beta1 "github.com/crossplane-contrib/provider-helm/apis/release/v1beta1"
+	crossplanev1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/vshn/appcat-service-postgresql/apis/conditions"
+	"github.com/vshn/appcat-service-postgresql/apis/postgresql/v1alpha1"
+	"github.com/vshn/appcat-service-postgresql/operator/helmvalues"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	// PostgresqlPodName the name of the postgresql instance pod running on the cluster
+	PostgresqlPodName string = "postgresql-0"
+)
+
+// UpdateStandalonePipeline is a pipeline that updates an existing instance in the target deployment namespace.
+type UpdateStandalonePipeline struct {
+	operatorNamespace   string
+	client              client.Client
+	instance            *v1alpha1.PostgresqlStandalone
+	config              *v1alpha1.PostgresqlStandaloneOperatorConfig
+	helmValues          helmvalues.V
+	deploymentNamespace *corev1.Namespace
+	helmRelease         *helmv1beta1.Release
+
+	connectionSecret      *corev1.Secret
+	persistentVolumeClaim *corev1.PersistentVolumeClaim
+}
+
+// NewUpdateStandalonePipeline creates a new pipeline with the required dependencies.
+func NewUpdateStandalonePipeline(client client.Client, instance *v1alpha1.PostgresqlStandalone, operatorNamespace string) *UpdateStandalonePipeline {
+	return &UpdateStandalonePipeline{
+		instance:          instance,
+		client:            client,
+		operatorNamespace: operatorNamespace,
+	}
+}
+
+// RunInitialUpdatePipeline executes the pipeline with configured business logic steps.
+// This should only be executed once per pipeline as it stores intermediate results in the struct.
+func (u *UpdateStandalonePipeline) RunInitialUpdatePipeline(ctx context.Context) error {
+	return pipeline.NewPipeline().
+		WithSteps(
+			pipeline.NewStepFromFunc("fetch operator config", u.fetchOperatorConfig),
+
+			pipeline.NewPipeline().WithNestedSteps("ensure connection secret",
+				pipeline.NewStepFromFunc("fetch connection secret", u.fetchConnectionSecret),
+				pipeline.If(u.isRootUserNotConsistentWithSecretData,
+					pipeline.NewStepFromFunc("ensure connection secret", u.ensureConnectionSecret),
+				),
+			),
+
+			pipeline.NewPipeline().WithNestedSteps("ensure persistent volume claim",
+				pipeline.NewStepFromFunc("fetch persistent volume claim", u.fetchPVC),
+				pipeline.If(u.isStorageUpdated,
+					pipeline.NewPipeline().WithNestedSteps("update helm release",
+						pipeline.NewStepFromFunc("ensure persistent volume claim", u.ensurePersistentVolumeClaim),
+						pipeline.NewStepFromFunc("restart postgresql pod", u.restartPostgresqlPod),
```

Comment:
I just had an idea for this step which makes restarting a pod much easier:

Instead of deleting the target pod, we'll add an annotation to the statefulset via a helm value `primary.podAnnotations` and set the annotation like `postgresql.appcat.vshn.io/storage-size: 8Gi" or whatever was passed. When we have updated the PVC with the new value, we'll just update the annotation in the helm release and Kubernetes will do the restart for us since the Statefulset spec has changed. (changing annotations should be possible)